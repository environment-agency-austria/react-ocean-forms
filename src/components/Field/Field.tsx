/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import * as React from 'react';

import { getDeepValue } from '../../utils';
import { IFormContext, TFieldValues } from '../FormContext';
import { IValidationArgs, IValidationState, withValidation } from '../withValidation';
import { IFieldComponentFieldProps, IFieldComponentMeta, IFieldProps, IValueMeta, TFieldValue } from './Field.types';

interface IContextMeta extends IValueMeta {
  defaultValue?: TFieldValue;
  externalValue?: TFieldValue;
}

interface IFieldState {
  touched: boolean;
  dirty: boolean;
  value: TFieldValue;
  contextMeta: IContextMeta;
}

interface IMetaChanges {
  defaultValue: boolean;
  externalValue: boolean;
  meta: boolean;
}

/**
 * Wrapper for input fields managed by
 * the form component
 */
export class BaseField extends React.Component<IFieldProps, IFieldState> {
  public static displayName: string = 'Field';

  // tslint:disable-next-line:typedef
  public static defaultProps = {
    onChange: (): void => undefined,
    onBlur: (): void => undefined,
    getDisplayValue: (value: TFieldValue): TFieldValue => value,
    getSubmitValue: (value: TFieldValue): TFieldValue => value,
  };

  constructor(props: IFieldProps) {
    super(props);

    const {
      fullName,
      label,
      context,
      validation: {
        update: updateValidation,
      },
    } = props;

    this.checkFormContext();

    context.registerField(
      fullName,
      {
        label,

        updateValidation,
        validate: this.validate,
        reset: this.reset,
        getValue: this.getValue,
      },
    );

    this.state = {
      touched: false,
      dirty: false,
      value: BaseField.callGetDisplayValue(props, ''),
      contextMeta: {
        disabled: false,
        plaintext: false,
      },
    };
  }

  /**
   * Changes the field value based on changes in Form.defaultValues,
   * Form.values, Field.defaultValue or Field.value. The Field props
   * always override the Form props. Changes in defaultValue will only
   * update the field value, if there is no values or value prop present
   * and if the field hasn't been touched.
   */
  public static getDerivedStateFromProps(nextProps: IFieldProps, prevState: IFieldState): IFieldState | null {
    const {
      context: {
        disabled: newDisabled,
        plaintext: newPlaintext,
      },
    } = nextProps;

    // Get the default value from Field.defaultValue or Form.defaultValues
    const defaultValue = BaseField.getDefaultValue(nextProps);
    // Get the "external" value from Field.value or Form.values
    const externalValue = BaseField.getExternalValue(nextProps);

    // Get changes in default or external value and meta information (plaintext, disabled)
    // Meta information must trigger a state change as well, because they are passed to
    // Field.getDisplayValue and could result in a different return value of that function
    const changes = BaseField.getPropChanges(nextProps, prevState, defaultValue, externalValue);
    // Get the new value based on the information above, or undefined if we don't need any
    // value changes.
    const propValue = BaseField.getPropValue(defaultValue, externalValue, changes);

    // Update the Field state if needed, also remember the current prop values so we can
    // detect changes in the next call of getDerivedStateFromProps.
    if (propValue !== undefined) {
      return ({
        value: BaseField.callGetDisplayValue(nextProps, propValue),
        touched: false,
        dirty: false,
        contextMeta: {
          defaultValue,
          externalValue,
          disabled: newDisabled,
          plaintext: newPlaintext,
        },
      });
    }

    return null;
  }

  /**
   * Correctly calls the getDisplayValue callback
   * @param props Field props
   * @param value Field value
   */
  private static callGetDisplayValue(props: IFieldProps, value: TFieldValue | undefined): TFieldValue {
    const { getDisplayValue, context } = props;

    return getDisplayValue(
      value || '',
      BaseField.getValueMeta(context),
    );
  }

  /**
   * Calculates the new state value based on the parameters - it will return
   * the externalValue if it exists and has changes, or the default value
   * otherwise. If no state update is needed it will return undefined. Meta
   * changes will trigger a state change as well.
   * @param defaultValue Default value
   * @param externalValue External value
   * @param changes Changes object
   */
  private static getPropValue(
    defaultValue: TFieldValue | undefined, externalValue: TFieldValue | undefined, changes: IMetaChanges,
  ): TFieldValue | undefined {
    const hasExternalValue = externalValue !== undefined;

    if (hasExternalValue && (changes.externalValue || changes.meta)) {
      return externalValue;
    }

    if (!hasExternalValue && (changes.defaultValue || changes.meta)) {
      return defaultValue;
    }

    return undefined;
  }

  /**
   * Detects changes in the defaultValue, the externalValue and in
   * context.disabled or context.plaintext (merged as meta prop)
   * @param props Field props
   * @param state Field state
   * @param defaultValue Default value
   * @param externalValue External value
   */
  private static getPropChanges(
    props: IFieldProps, state: IFieldState, defaultValue: TFieldValue | undefined, externalValue: TFieldValue | undefined,
  ): IMetaChanges {
    const {
      context: {
        disabled: newDisabled,
        plaintext: newPlaintext,
      },
    } = props;

    const {
      contextMeta: {
        defaultValue: oldDefaultValue,
        externalValue: oldExternalValue,
        disabled: oldDisabled,
        plaintext: oldPlaintext,
      },
      touched,
    } = state;

    return {
      defaultValue: !touched && defaultValue !== oldDefaultValue,
      externalValue: externalValue !== oldExternalValue,
      meta: newDisabled !== oldDisabled || newPlaintext !== oldPlaintext,
    };
  }

  /**
   * Returns the local value if existing, otherwise tries to
   * extract the correct context value based on the fullName.
   * @param localValue Local value
   * @param contextValue Context value
   * @param fullName Field.fullName
   */
  private static getLocalOverridenValue(
    localValue: TFieldValue | undefined, contextValue: TFieldValues | undefined, fullName: string,
  ): TFieldValue | undefined {
    return localValue || getDeepValue(fullName, contextValue);
  }

  /**
   * Returns either the Field.value or the correct value from
   * Form.values
   * @param props Field props
   */
  private static getExternalValue(props: IFieldProps): TFieldValue | undefined {
    const { fullName, context: { values }, value } = props;

    return BaseField.getLocalOverridenValue(value, values, fullName);
  }

  /**
   * Returns either the Field.defaultValue or the correct default value
   * from Form.defaultValues
   * @param props Field props
   */
  private static getDefaultValue(props: IFieldProps): TFieldValue | undefined {
    const { fullName, context: { defaultValues }, defaultValue } = props;

    return BaseField.getLocalOverridenValue(defaultValue, defaultValues, fullName);
  }

  /**
   * Returns a meta object for the value lifecycle hooks
   * @param context Form context
   */
  private static getValueMeta(context: IFormContext): IValueMeta {
    return {
      disabled: context.disabled,
      plaintext: context.plaintext,
    };
  }

  /**
   * Unregisters the field from the form
   */
  public componentWillUnmount(): void {
    const { context, fullName } = this.props;
    context.unregisterField(fullName);
  }

  /**
   * Returns the current field value
   */
  private getValue = (): TFieldValue => {
    const { value } = this.state;
    const { getSubmitValue, context } = this.props;

    return getSubmitValue(
      value,
      BaseField.getValueMeta(context),
    );
  }

  /**
   * Returns the correct asyncValidateOnChange setting,
   * where the field setting takes priorty over the
   * form setting
   */
  private getAsyncValidateOnChangeSetting(): boolean {
    const {
      asyncValidateOnChange: propChange,
      context: { asyncValidateOnChange: contextChange },
    } = this.props;

    return propChange === undefined ? contextChange : propChange;
  }

  /**
   * Checks if the Field is inside a valid form context
   * and throws an user friendly error if not
   */
  private checkFormContext(): void {
    const { context, fullName } = this.props;
    if (!context || typeof context.registerField !== 'function') {
      throw new Error(
        `Could not find a form context for field "${fullName}". `
        + 'Fields can only be used inside a Form tag.',
      );
    }
  }

  /**
   * Resets the field to its default state
   */
  private reset = (): void => {
    const { contextMeta: { externalValue, defaultValue } } = this.state;
    const { validation, onChange } = this.props;

    const value = BaseField.callGetDisplayValue(
      this.props,
      externalValue || defaultValue,
    );

    this.setState({
      value,
      touched: false,
      dirty: false,
    });

    validation.reset();
    onChange(value);
  }

  /**
   * Validates the field
   * @param args Validation arguments
   */
  private validate = async (args?: Partial<IValidationArgs>): Promise<IValidationState> => {
    const { value } = this.state;
    const { validation: { validate }, getSubmitValue, context } = this.props;

    return validate(
      getSubmitValue(
        value,
        BaseField.getValueMeta(context),
      ),
      args,
    );
  }

  /**
   * Handles the change event of an input field -
   * triggers any validation if needed and updates
   * the field state accordingly.
   * @param event Event object
   */
  private handleFieldChanged = (event: React.ChangeEvent<HTMLInputElement>): void => {
    const { value } = event.target;
    const {
      fullName,
      context,
      validation: { validate },
      getSubmitValue,
      onChange,
    } = this.props;

    this.setState({
      value,
      dirty: true,
      touched: true,
    });

    const asyncValidateOnChange = this.getAsyncValidateOnChangeSetting();
    const submitValue = getSubmitValue(
      value,
      BaseField.getValueMeta(context),
    );

    validate(
      submitValue,
      { checkAsync: asyncValidateOnChange },
    );

    context.notifyFieldEvent(fullName, 'change', submitValue);
    onChange(submitValue);
  }

  /**
   * Handles the blur event of an input field -
   * triggers any validation if needed and updates
   * the field state accordingly.
   */
  private handleFieldBlurred = (): void => {
    const {
      fullName,
      validation: { validate },
      context,
      onBlur,
      getSubmitValue,
    } = this.props;
    const { value, dirty } = this.state;

    const asyncValidateOnChange = this.getAsyncValidateOnChangeSetting();
    const submitValue = getSubmitValue(
      value,
      BaseField.getValueMeta(context),
    );

    if (dirty && !asyncValidateOnChange) { validate(submitValue); }
    context.notifyFieldEvent(fullName, 'blur');
    onBlur();
  }

  /**
   * Creates the properties that can be directly
   * mapped to the input component (e.g. html input)
   */
  private createFieldProps(): IFieldComponentFieldProps {
    const { context: { disabled }, fullName } = this.props;
    const { value } = this.state;

    return {
      value,
      disabled,
      id: fullName,
      name: fullName,
      onChange: this.handleFieldChanged,
      onBlur: this.handleFieldBlurred,
    };
  }

  /**
   * Creates the meta properties with meta information
   * used by the input component
   */
  private createMetaProps(): IFieldComponentMeta {
    const {
      context: {
        stringFormatter,
        plaintext,
      },
      validation: {
        valid,
        error,
        isValidating,
      },
    } = this.props;

    const { touched } = this.state;

    return {
      valid,
      error,
      isValidating,
      touched,
      stringFormatter,
      plaintext,
    };
  }

  // tslint:disable-next-line:member-ordering
  public render(): JSX.Element {
    const {
      // tslint:disable-next-line:naming-convention
      component: InputComponent,
      context,
      fullName,
      validation,
      onChange,
      onBlur,
      defaultValue,
      value,
      name,
      asyncValidateOnChange,
      getDisplayValue,
      getSubmitValue,
      ...attributes
    } = this.props;

    const field = this.createFieldProps();
    const meta = this.createMetaProps();

    return (
      <InputComponent
        {...attributes}
        field={field}
        meta={meta}
      />
    );
  }
}

export const Field = withValidation(BaseField);
