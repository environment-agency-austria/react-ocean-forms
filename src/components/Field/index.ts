export * from './Field';
export * from './Field.types';
