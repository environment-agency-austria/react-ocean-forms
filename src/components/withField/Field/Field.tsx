/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';

import { getDeepValue } from '../../../utils';
import { IFieldValues, IFormContext } from '../../FormContext';
import { IValidationArgs, IValidationState, withValidation } from '../../withValidation';
import {
  IFieldChangedEvent,
  IFieldComponentFieldProps,
  IFieldComponentMeta,
  IFieldProps,
  IValueMeta,
  TBasicFieldValue,
} from './Field.types';

interface IContextMeta extends IValueMeta {
  defaultValue?: TBasicFieldValue;
  externalValue?: TBasicFieldValue;
}

interface IFieldState {
  touched: boolean;
  dirty: boolean;
  value: TBasicFieldValue;
  contextMeta: IContextMeta;
}

interface IMetaChanges {
  defaultValue: boolean;
  externalValue: boolean;
  meta: boolean;
}

/**
 * Wrapper for input fields managed by
 * the form component
 */
export class BaseField extends React.Component<IFieldProps, IFieldState> {
  public static displayName: string = 'Field';

  constructor(props: IFieldProps) {
    super(props);

    const {
      fullName,
      label,
      context,
      validation: {
        update: updateValidation,
      },
    } = props;

    this.checkFormContext();

    context.registerField(
      fullName,
      {
        label,

        updateValidation,
        validate: this.validate,
        reset: this.reset,
        getValue: this.getValue,
      },
    );

    this.state = {
      touched: false,
      dirty: false,
      value: BaseField.callGetDisplayValue(props, ''),
      contextMeta: {
        disabled: false,
        plaintext: false,
      },
    };
  }

  /**
   * Changes the field value based on changes in Form.defaultValues,
   * Form.values, Field.defaultValue or Field.value. The Field props
   * always override the Form props. Changes in defaultValue will only
   * update the field value, if there is no values or value prop present
   * and if the field hasn't been touched.
   */
  public static getDerivedStateFromProps(nextProps: IFieldProps, prevState: IFieldState): IFieldState | null {
    const {
      context: {
        disabled: newDisabled,
        plaintext: newPlaintext,
      },
    } = nextProps;

    // Get the default value from Field.defaultValue or Form.defaultValues
    const defaultValue = BaseField.getDefaultValue(nextProps);
    // Get the "external" value from Field.value or Form.values
    const externalValue = BaseField.getExternalValue(nextProps);

    // Get changes in default or external value and meta information (plaintext, disabled)
    // Meta information must trigger a state change as well, because they are passed to
    // Field.getDisplayValue and could result in a different return value of that function
    const changes = BaseField.getPropChanges(nextProps, prevState, defaultValue, externalValue);
    // Get the new value based on the information above, or undefined if we don't need any
    // value changes.
    const propValue = BaseField.getPropValue(defaultValue, externalValue, changes);

    // Update the Field state if needed, also remember the current prop values so we can
    // detect changes in the next call of getDerivedStateFromProps.
    if (propValue !== undefined) {
      return ({
        value: BaseField.callGetDisplayValue(nextProps, propValue),
        touched: false,
        dirty: false,
        contextMeta: {
          defaultValue,
          externalValue,
          disabled: newDisabled,
          plaintext: newPlaintext,
        },
      });
    }

    return null;
  }

  /**
   * Correctly calls the getDisplayValue callback
   * @param props Field props
   * @param value Field value
   */
  private static callGetDisplayValue(props: IFieldProps, value: TBasicFieldValue | undefined): TBasicFieldValue {
    const {
      context,
      getDisplayValue = (val: TBasicFieldValue): TBasicFieldValue => val,
    } = props;

    return getDisplayValue(
      value === undefined ? '' : value,
      BaseField.getValueMeta(context),
    );
  }

  /**
   * Calculates the new state value based on the parameters - it will return
   * the externalValue if it exists and has changes, or the default value
   * otherwise. If no state update is needed it will return undefined. Meta
   * changes will trigger a state change as well.
   * @param defaultValue Default value
   * @param externalValue External value
   * @param changes Changes object
   */
  private static getPropValue(
    defaultValue: TBasicFieldValue | undefined, externalValue: TBasicFieldValue | undefined, changes: IMetaChanges,
  ): TBasicFieldValue | undefined {
    const hasExternalValue = externalValue !== undefined;

    if (hasExternalValue && (changes.externalValue || changes.meta)) {
      return externalValue;
    }

    if (!hasExternalValue && (changes.defaultValue || changes.meta)) {
      return defaultValue;
    }

    return undefined;
  }

  /**
   * Detects changes in the defaultValue, the externalValue and in
   * context.disabled or context.plaintext (merged as meta prop)
   * @param props Field props
   * @param state Field state
   * @param defaultValue Default value
   * @param externalValue External value
   */
  private static getPropChanges(
    props: IFieldProps, state: IFieldState, defaultValue: TBasicFieldValue | undefined, externalValue: TBasicFieldValue | undefined,
  ): IMetaChanges {
    const {
      context: {
        disabled: newDisabled,
        plaintext: newPlaintext,
      },
    } = props;

    const {
      contextMeta: {
        defaultValue: oldDefaultValue,
        externalValue: oldExternalValue,
        disabled: oldDisabled,
        plaintext: oldPlaintext,
      },
      touched,
    } = state;

    return {
      defaultValue: !touched && defaultValue !== oldDefaultValue,
      externalValue: externalValue !== oldExternalValue,
      meta: newDisabled !== oldDisabled || newPlaintext !== oldPlaintext,
    };
  }

  /**
   * Returns the local value if existing, otherwise tries to
   * extract the correct context value based on the fullName.
   * @param localValue Local value
   * @param contextValue Context value
   * @param fullName Field.fullName
   */
  private static getLocalOverridenValue(
    localValue: TBasicFieldValue | undefined, contextValue: Partial<IFieldValues> | undefined, fullName: string,
  ): TBasicFieldValue | undefined {
    return localValue === undefined ? getDeepValue(fullName, contextValue) : localValue;
  }

  /**
   * Returns either the Field.value or the correct value from
   * Form.values
   * @param props Field props
   */
  private static getExternalValue(props: IFieldProps): TBasicFieldValue | undefined {
    const { fullName, context: { values }, value } = props;

    return BaseField.getLocalOverridenValue(value, values, fullName);
  }

  /**
   * Returns either the Field.defaultValue or the correct default value
   * from Form.defaultValues
   * @param props Field props
   */
  private static getDefaultValue(props: IFieldProps): TBasicFieldValue | undefined {
    const { fullName, context: { defaultValues }, defaultValue } = props;

    return BaseField.getLocalOverridenValue(defaultValue, defaultValues, fullName);
  }

  /**
   * Returns a meta object for the value lifecycle hooks
   * @param context Form context
   */
  private static getValueMeta(context: IFormContext): IValueMeta {
    return {
      disabled: context.disabled,
      plaintext: context.plaintext,
    };
  }

  /**
   * Unregisters the field from the form
   */
  public componentWillUnmount(): void {
    const { context, fullName } = this.props;
    context.unregisterField(fullName);
  }

  /**
   * Returns the current field value
   */
  private getValue = (): TBasicFieldValue => {
    const { value } = this.state;

    return this.getSubmitValue(value);
  }

  private getSubmitValue = (value: TBasicFieldValue): TBasicFieldValue => {
    const {
      context,
      getSubmitValue = (val: TBasicFieldValue): TBasicFieldValue => val,
    } = this.props;

    return getSubmitValue(
      value,
      BaseField.getValueMeta(context),
    );
  }

  /**
   * Returns the correct asyncValidateOnChange setting,
   * where the field setting takes priorty over the
   * form setting
   */
  private getAsyncValidateOnChangeSetting(): boolean {
    const {
      asyncValidateOnChange: propChange,
      context: { asyncValidateOnChange: contextChange },
    } = this.props;

    return propChange === undefined ? contextChange : propChange;
  }

  /**
   * Checks if the Field is inside a valid form context
   * and throws an user friendly error if not
   */
  private checkFormContext(): void {
    const { context, fullName } = this.props;
    if (context === undefined || typeof context.registerField !== 'function') {
      throw new Error(
        `Could not find a form context for field "${fullName}". `
        + 'Fields can only be used inside a Form tag.',
      );
    }
  }

  /**
   * Resets the field to its default state
   */
  private reset = (): void => {
    const { contextMeta: { externalValue, defaultValue } } = this.state;
    const { validation, onChange } = this.props;

    const value = BaseField.callGetDisplayValue(
      this.props,
      externalValue === undefined ? defaultValue : externalValue,
    );

    this.setState({
      value,
      touched: false,
      dirty: false,
    });

    validation.reset();
    if (onChange !== undefined) { onChange(value); }
  }

  /**
   * Validates the field
   * @param args Validation arguments
   */
  private validate = async (args?: Partial<IValidationArgs>): Promise<IValidationState> => {
    const { value } = this.state;
    const { validation: { validate } } = this.props;

    return validate(
      this.getSubmitValue(value),
      args,
    );
  }

  /**
   * Handles the change event of an input field -
   * triggers any validation if needed and updates
   * the field state accordingly.
   * @param event Event object
   */
  private handleFieldChanged = (event: IFieldChangedEvent): void => {
    const { value } = event.target;
    const {
      fullName,
      context,
      validation: { validate },
      onChange,
    } = this.props;

    this.setState({
      value,
      dirty: true,
      touched: true,
    });

    const asyncValidateOnChange = this.getAsyncValidateOnChangeSetting();
    const submitValue = this.getSubmitValue(value);

    void validate(
      submitValue,
      { checkAsync: asyncValidateOnChange },
    );

    context.notifyFieldEvent(fullName, 'change', submitValue);
    if (onChange !== undefined) { onChange(submitValue); }
  }

  /**
   * Handles the blur event of an input field -
   * triggers any validation if needed and updates
   * the field state accordingly.
   */
  private handleFieldBlurred = (): void => {
    const {
      fullName,
      validation: { validate },
      context,
      onBlur,
    } = this.props;
    const { value, dirty } = this.state;

    const asyncValidateOnChange = this.getAsyncValidateOnChangeSetting();
    const submitValue = this.getSubmitValue(value);

    if (dirty && !asyncValidateOnChange) { void validate(submitValue); }
    context.notifyFieldEvent(fullName, 'blur');
    if (onBlur !== undefined) { onBlur(); }
  }

  /**
   * Creates the properties that can be directly
   * mapped to the input component (e.g. html input)
   */
  private createFieldProps(): IFieldComponentFieldProps {
    const { context: { disabled }, fullName } = this.props;
    const { value } = this.state;

    return {
      value,
      disabled,
      id: fullName,
      name: fullName,
      onChange: this.handleFieldChanged,
      onBlur: this.handleFieldBlurred,
    };
  }

  /**
   * Creates the meta properties with meta information
   * used by the input component
   */
  private createMetaProps(): IFieldComponentMeta {
    const {
      context: {
        stringFormatter,
        plaintext,
      },
      validation: {
        valid,
        error,
        isValidating,
        isRequired,
      },
    } = this.props;

    const { touched } = this.state;

    return {
      valid,
      error,
      isValidating,
      isRequired,
      touched,
      stringFormatter,
      plaintext,
    };
  }

  // tslint:disable-next-line:member-ordering
  public render(): JSX.Element {
    const { render } = this.props;

    return render(
      this.createFieldProps(),
      this.createMetaProps(),
    );
  }
}

export const Field = withValidation(BaseField);
