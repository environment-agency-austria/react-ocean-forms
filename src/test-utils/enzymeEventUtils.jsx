/**
 * Defines some often used keycodes for
 * event testing
 */
export const KEYCODE = {
  BACKSPACE: 8,
  TAB: 9,
  ENTER: 13,
  ESCAPE: 27,
  DOWN_ARROW: 40,
};

/**
 * Generates a mocked event to be used with
 * enzymes simulate in situations where the
 * tested code wants to call preventDefault
 * @param {object} props Additional event props
 */
export const mockEvent = props => ({
  preventDefault: () => {},
  ...props,
});

export default mockEvent;
