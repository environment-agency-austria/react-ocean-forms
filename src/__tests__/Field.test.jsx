import React from 'react';
import { shallow } from 'enzyme';

import { createMockFormContext, createMockValidation } from '../test-utils/enzymeFormContext';
import { BaseField } from '../Field';

describe('<Field />', () => {
  const TestComponent = () => (<div id="test-component" />);

  let fieldState = null;
  const registerCallback = (name, state) => { fieldState = state; };

  const fieldName = 'unitField';
  const fullName = fieldName;
  const fieldLabel = 'Unit field';
  const formContext = createMockFormContext(registerCallback);
  const validation = createMockValidation();

  const onChangeHandler = jest.fn();
  const onBlurHandler = jest.fn();

  const getDisplayValue = jest.fn().mockImplementation(value => value);
  const getSubmitValue = jest.fn().mockImplementation(value => value);

  const setup = props => shallow((
    <BaseField
      name={fieldName}
      fullName={fullName}
      label={fieldLabel}
      component={TestComponent}
      context={formContext}
      validation={validation}
      onChange={onChangeHandler}
      onBlur={onBlurHandler}
      getDisplayValue={getDisplayValue}
      getSubmitValue={getSubmitValue}
      {...props}
    />
  ));
  const wrapper = setup();

  it('should render without error', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call the getDisplayValue function', () => {
    expect(getDisplayValue).toHaveBeenLastCalledWith(
      '',
      { disabled: false, plaintext: false },
    );
  });

  it('should register itself in the form context', () => {
    expect(formContext.registerField).toHaveBeenCalledWith(
      fieldName,
      {
        label: fieldLabel,
        validate: expect.any(Function),
        reset: expect.any(Function),
        getValue: expect.any(Function),
        updateValidation: validation.update,
      },
    );
  });

  const changeValue = 'blubb';

  describe('onChange event handling', () => {
    const triggerChange = () => {
      const inputElement = wrapper.find('TestComponent');
      inputElement.prop('field').onChange({
        target: {
          name: fieldName,
          value: changeValue,
        },
      });
      wrapper.update();
    };

    it('should remember the changed value', () => {
      triggerChange();
      expect(wrapper.find('TestComponent').prop('field').value).toBe(changeValue);
    });

    it('should call the getSubmitValue function', () => {
      expect(getSubmitValue).toHaveBeenLastCalledWith(
        changeValue,
        { disabled: false, plaintext: false },
      );
    });

    it('should call the validate function', () => {
      expect(validation.validate).toHaveBeenLastCalledWith(
        changeValue,
        { checkAsync: false },
      );
    });

    it('should respect the Form.asyncValidateOnChange configuration', () => {
      formContext.asyncValidateOnChange = true;
      wrapper.setProps({ context: formContext });

      triggerChange();
      expect(validation.validate).toHaveBeenLastCalledWith(
        changeValue,
        { checkAsync: true },
      );
    });

    it('should respect the Field.asyncValidateOnChange configuration', () => {
      formContext.asyncValidateOnChange = false;
      wrapper.setProps({ context: formContext, asyncValidateOnChange: true });

      triggerChange();
      expect(validation.validate).toHaveBeenLastCalledWith(
        changeValue,
        { checkAsync: true },
      );

      wrapper.setProps({ asyncValidateOnChange: false });
    });

    it('should notify the form context', () => {
      expect(formContext.notifyFieldEvent).toHaveBeenLastCalledWith(
        fieldName,
        'change',
        changeValue,
      );
    });

    it('should call the onChange handler', () => {
      expect(onChangeHandler).toHaveBeenLastCalledWith(changeValue);
    });

    it('should not crash without an onChange handler', () => {
      wrapper.setProps({ onChange: undefined });
      expect(() => triggerChange()).not.toThrowError();

      wrapper.setProps({ onChange: onChangeHandler });
      onChangeHandler.mockReset();
    });
  });

  describe('onBlur event handling', () => {
    const triggerBlur = () => {
      const inputElement = wrapper.find('TestComponent');
      inputElement.prop('field').onBlur();
      wrapper.update();
    };

    const resetMocks = () => {
      validation.validate.mockReset();
      formContext.notifyFieldEvent.mockReset();
    };

    beforeAll(() => {
      resetMocks();
      triggerBlur();
    });

    afterAll(() => {
      resetMocks();
    });

    it('should call the getSubmitValue function', () => {
      expect(getSubmitValue).toHaveBeenLastCalledWith(
        changeValue,
        { disabled: false, plaintext: false },
      );
    });

    it('should call the validate function', () => {
      expect(validation.validate).toHaveBeenLastCalledWith(changeValue);
    });

    it('should notify the form context', () => {
      expect(formContext.notifyFieldEvent).toHaveBeenLastCalledWith(
        fieldName,
        'blur',
      );
    });

    it('should call the onBlur handler', () => {
      expect(onBlurHandler).toHaveBeenCalled();
    });

    it('should not crash without an onBlur handler', () => {
      wrapper.setProps({ onBlur: undefined });
      expect(() => triggerBlur()).not.toThrowError();

      wrapper.setProps({ onBlur: onBlurHandler });
      onBlurHandler.mockReset();
    });

    it('should not validate if the state is not dirty', () => {
      resetMocks();
      wrapper.setState({ dirty: false });
      wrapper.update();
      triggerBlur();

      expect(validation.validate).not.toHaveBeenCalled();
    });
  });

  describe('form context callbacks', () => {
    it('should return the correct value', () => {
      expect(fieldState.getValue()).toBe(changeValue);
    });

    it('should correctly pass the validate method', () => {
      const validateArgs = { foo: 'bar' };

      fieldState.validate(validateArgs);
      expect(validation.validate).toHaveBeenLastCalledWith(changeValue, validateArgs);
    });

    describe('reset', () => {
      it('should correctly reset to its defaultValue', () => {
        fieldState.reset();
        wrapper.update();

        const inputElement = wrapper.find('TestComponent');
        expect(inputElement.prop('field').value).toBe('');

        expect(validation.reset).toHaveBeenCalled();
      });

      it('should call the onChange handler', () => {
        expect(onChangeHandler).toHaveBeenLastCalledWith('');
      });
    });
  });

  describe('value callbacks', () => {
    const get42 = jest.fn().mockImplementation(() => 42);
    const get84 = jest.fn().mockImplementation(() => 84);

    beforeAll(() => {
      formContext.defaultValues = { [fieldName]: 33 };
      wrapper.setProps({
        getDisplayValue: get42,
        getSubmitValue: get84,
        context: formContext,
      });
      wrapper.update();
    });

    afterAll(() => {
      wrapper.setProps({ getDisplayValue, getSubmitValue });
    });

    it('should display 42 as value', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should return 84 as value', () => {
      const fieldValue = wrapper.instance().getValue();
      expect(get84).toHaveBeenLastCalledWith(42, { disabled: false, plaintext: false });
      expect(fieldValue).toBe(84);
    });
  });

  it('should work without value callbacks', () => {
    wrapper.setProps({
      getDisplayValue: undefined,
      getSubmitValue: undefined,
    });
    wrapper.update();

    expect(() => {
      wrapper.instance().getValue();
    }).not.toThrowError();
    expect(wrapper).toMatchSnapshot();
  });

  it('should set its value to the defaultValue', () => {
    const newDefaultValue = 'foobar';
    formContext.defaultValues = { [fieldName]: newDefaultValue };
    wrapper.setProps({ context: formContext });
    wrapper.update();

    const inputElement = wrapper.find('TestComponent');
    expect(inputElement.prop('field').value).toBe(newDefaultValue);
  });

  it('should use an empty string if there is no defaultValue', () => {
    formContext.defaultValues = { foo: 'bar' };
    wrapper.setProps({ context: formContext });
    wrapper.update();

    const inputElement = wrapper.find('TestComponent');
    expect(inputElement.prop('field').value).toBe('');
  });

  it('should correctly unregister on unmount', () => {
    wrapper.unmount();
    expect(formContext.unregisterField).toHaveBeenCalledWith(fieldName);
  });
});
