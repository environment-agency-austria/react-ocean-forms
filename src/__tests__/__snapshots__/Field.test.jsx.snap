// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Field /> should render without error 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Field
    asyncValidateOnChange={null}
    component={[Function]}
    context={
      Object {
        "asyncValidateOnChange": false,
        "asyncValidationWait": 400,
        "busy": false,
        "defaultValues": Object {},
        "disabled": false,
        "fieldPrefix": null,
        "getFieldState": [MockFunction],
        "getValues": [MockFunction],
        "notifyFieldEvent": [MockFunction],
        "registerField": [MockFunction] {
          "calls": Array [
            Array [
              "unitField",
              Object {
                "getValue": [Function],
                "label": "Unit field",
                "reset": [Function],
                "updateValidation": [MockFunction],
                "validate": [Function],
              },
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        },
        "registerListener": [MockFunction],
        "stringFormatter": [MockFunction],
        "unregisterField": [MockFunction],
        "unregisterListener": [MockFunction],
      }
    }
    fullName="unitField"
    label="Unit field"
    name="unitField"
    validation={
      Object {
        "error": null,
        "isValidating": false,
        "reset": [MockFunction],
        "update": [MockFunction],
        "valid": true,
        "validate": [MockFunction],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "asyncValidateOnChange": null,
      "field": Object {
        "disabled": false,
        "id": "unitField",
        "name": "unitField",
        "onBlur": [Function],
        "onChange": [Function],
        "value": "",
      },
      "label": "Unit field",
      "meta": Object {
        "error": null,
        "isValidating": false,
        "stringFormatter": [MockFunction],
        "touched": false,
        "valid": true,
      },
      "name": "unitField",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "asyncValidateOnChange": null,
        "field": Object {
          "disabled": false,
          "id": "unitField",
          "name": "unitField",
          "onBlur": [Function],
          "onChange": [Function],
          "value": "",
        },
        "label": "Unit field",
        "meta": Object {
          "error": null,
          "isValidating": false,
          "stringFormatter": [MockFunction],
          "touched": false,
          "valid": true,
        },
        "name": "unitField",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
